### PR TITLE
State index ledger database

### PR DESCRIPTION
**Pull request by sideninja**
Closes: #449 

## Description

This is adding a ledger database implementation for creating a local state. This is the first version and it might be changing.

______

For contributor use:

- [ ] Targeted PR against `master` branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [standards mentioned here](https://github.com/onflow/flow-nft/blob/master/CONTRIBUTING.md#styleguides).
- [ ] Updated relevant documentation 
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Added appropriate labels 

## Summary by CodeRabbit

- **New Features**
    - Added a new field `PrecompiledCalls` to enhance transaction receipt data.
    - Introduced constants for improved ledger management.
    - Launched a new `Ledger` implementation for better key-value operations.

- **Improvements**
    - Simplified function interfaces by removing unnecessary parameters, enhancing maintainability.
    - Enhanced transaction and block ingestion tests for better clarity and error handling.

- **Bug Fixes**
    - Improved safety checks for receipt storage to prevent block height mismatches.

- **Tests**
    - Updated test suite to reflect changes in method signatures and added scenarios for error conditions.

### CODE REVIEW

File: storage/index_testsuite.go [hunk 1]
```diff
@@ -196,12 +196,22 @@ func (s *ReceiptTestSuite) TestStoreReceipt() {
...
 	})
+
+	s.Run("fail to store multiple receipts with different heights", func() {
+		receipts := []*models.StorageReceipt{
+			mocks.NewReceipt(1, common.HexToHash("0x1")),
+			mocks.NewReceipt(2, common.HexToHash("0x2")),
+		}
+
+		err := s.ReceiptIndexer.Store(receipts, nil)
+		s.Require().EqualError(err, "can't store receipts for multiple heights")
```
Comment (m-Peter) [inline storage/index_testsuite.go:207]: Nice :ok_hand:

File: models/receipt.go [hunk 2]
```diff
@@ -31,6 +31,7 @@ type StorageReceipt struct {
 	BlockNumber       *big.Int         `json:"blockNumber,omitempty"`
 	TransactionIndex  uint             `json:"transactionIndex"`
 	RevertReason      []byte           `json:"revertReason"`
+	PrecompiledCalls  []byte
```
Comment (m-Peter) [inline models/receipt.go:34]: :warning: Reminder to re-index, upon deploying a release that contains this change :warning:
You will also need to set this value, upon decoding the `EVM.TransactionExecuted` event, here: https://github.com/onflow/flow-evm-gateway/blob/main/models/transaction.go#L200

Comment (sideninja) [reply models/receipt.go:34]: Updated

**Review by m-Peter — APPROVED** [review]
Awesome :tada:

**Pull request merged**